File: apps/desktop/src/main/fs/workspace/writtenScan.ts
/**
 * 产物写回 · 本地执行腿：扫描「本次执行改了工作区哪些文件」。
 *
 * 服务端 `tools/sandbox/written_scan.py` 的镜像实现（同一口径，逐条对应；改一侧须同步
 * 另一侧）。云端 gVisor 与本机一样 bind 落盘，改动靠事后扫描；本机执行（本模块）与
 * sidecar 直接以工作区为 cwd 让脚本写盘，不事后看盘就没人能如实填
 * `ExecutionResult.written_files` —— 桌面用户跑脚本产出的文件于是全部不进交付物台账
 * （不上交付 artifacts、不进用户面路径、CEO 清单也看不见）。
 *
 * 为什么是「执行后单次有界扫描 + mtime 截止」，而不是执行前后各拍一次全树快照：
 *
 * - 成本几乎全在**目录枚举**次数上，与文件数关系不大。实测一个大仓（剪掉 node_modules /
 *   .git 后仍有 19.5k 目录 / 196k 文件）：整树约 1 秒，只枚举目录不 stat 也要 0.9 秒。
 *   前后两次快照 = 双倍，会把一条 100ms 的脚本拖成两秒。
 * - 有界：从 cwd 起 **BFS**，撞到 {@link SCAN_MAX_DIRS} 或 {@link SCAN_BUDGET_MS} 立刻
 *   收手。BFS 保证被截断的永远是最深的尾巴，而产物几乎都落在 cwd 或浅层子目录；于是
 *   「超大工作区」退化成「浅层如实、极深处可能漏报」，而不是每次执行整体变慢。
 *
 * 排除面复用 `workspaceIgnore`（不另起第三份名单）：路径感知旁路区
 * `AgentCore/{index,trash,baselines}`、系统噪音目录、系统噪音后缀（`*.db` / `*.pyc`）、
 * 符号链接。**不**排除 AI 噪音后缀——AI 生成的 `.png` 图表正是交付物。
 */
import { promises as fs } from "node:fs";
import { join, relative } from "node:path";
import type { WorkspaceOpResult } from "@shared/ipc-contract";
import {
  shouldSkipDirName,
  shouldSkipSystemFileName,
} from "../workspaceIgnore";
import { toPosix } from "./result";

/** 目录枚举上限（实测约 45µs/目录，4000 目录 ≈ 180ms，与墙钟预算同量级）。 */
export const SCAN_MAX_DIRS = 4000;
/** 墙钟兜底：网络盘 / 被杀毒软件挂钩的目录上单次 readdir 可能慢几个数量级。 */
export const SCAN_BUDGET_MS = 250;
/** 报告条数上限（与服务端 `_MAX_FILES` 取齐）。 */
export const SCAN_MAX_FILES = 200;

/**
 * mtime 截止余量。文件系统时间戳粒度比墙钟粗且**向下截断**：刚起步就写的文件，mtime
 * 可能落在启动时刻之前一点（实测 Windows/NTFS ~1ms；Linux 粗粒度时钟 ≤10ms）。让
 * 100ms 即十倍余量，何况解释器启动本身就要几十毫秒。
 *
 * 不让更多：余量就是误报窗口（窗口内**别人**刚改过的文件会被算进本次执行）。已知不覆盖
 * FAT/exFAT（修改时间 2 秒粒度）与时钟落后的网络盘——宁可如实说明，不用秒级余量把所有
 * 人的精度赔进去。执行前就存在且未变动的文件在哪种盘上都不会命中（mtime 是旧的）。
 */
export const SCAN_MTIME_MARGIN_MS = 100;

export function writtenScanCutoffMs(): number {
  return Date.now() - SCAN_MTIME_MARGIN_MS;
}

export type WrittenScan = { files: string[]; truncated: boolean };

/**
 * `rootAbs` 下 mtime 不早于 `cutoffMs` 的常规文件（相对 `rootAbs` 的 POSIX 路径）。
 *
 * 单条目的读取失败（权限 / 占用 / 竞态删除）只跳过该条目 —— 执行本身已经成功了，
 * 记账扫描不得把它变成失败。
 */
export async function scanWrittenFiles(
  rootAbs: string,
  cutoffMs: number,
  opts?: { maxDirs?: number; budgetMs?: number; maxFiles?: number },
): Promise<WrittenScan> {
  const maxDirs = opts?.maxDirs ?? SCAN_MAX_DIRS;
  const maxFiles = opts?.maxFiles ?? SCAN_MAX_FILES;
  const deadline = Date.now() + (opts?.budgetMs ?? SCAN_BUDGET_MS);
  const queue: Array<{ abs: string; rel: string }> = [
    { abs: rootAbs, rel: "" },
  ];
  const found: string[] = [];
  let dirsSeen = 0;
  let truncated = false;

  while (queue.length > 0) {
    if (dirsSeen >= maxDirs || Date.now() >= deadline) {
      truncated = true;
      break;
    }
    const cur = queue.shift();
    if (!cur) break;
    dirsSeen += 1;
    let dirents: import("node:fs").Dirent[];
    try {
      dirents = await fs.readdir(cur.abs, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const d of dirents) {
      // 符号链接既不下潜也不上报（与服务端 staging 写回腿同口径）。
      if (d.isSymbolicLink()) continue;
      const childRel = cur.rel ? `${cur.rel}/${d.name}` : d.name;
      if (d.isDirectory()) {
        if (shouldSkipDirName(d.name, cur.rel)) continue;
        queue.push({ abs: join(cur.abs, d.name), rel: childRel });
        continue;
      }
      if (!d.isFile()) continue;
      if (shouldSkipSystemFileName(d.name)) continue;
      try {
        const st = await fs.lstat(join(cur.abs, d.name));
        if (st.mtimeMs >= cutoffMs) found.push(childRel);
      } catch {
        // 竞态删除 / 权限拒绝：漏掉这一条，不影响其余。
      }
    }
    if (found.length >= maxFiles) {
      truncated = true;
      break;
    }
  }

  found.sort();
  return { files: found.slice(0, maxFiles), truncated };
}

/** 超时 / 起不来的信封统一用 `exit_code: -1`（见 `exec.ts` timeoutResult）。 */
function completedOnItsOwn(value: Record<string, unknown>): boolean {
  return Number(value.exit_code) !== -1;
}

/**
 * 给一次 `execute` 的成功信封补上 `written_files`（工作区相对 POSIX 路径）。
 *
 * 只在进程**自己跑完**（任何退出码）时报：与云端 copy-out 腿同一条规则。云端跳过是因为
 * 强杀后压根没落盘；本机的写其实已经在盘上了，但把强杀可能写了一半的文件当产物广告出去
 * 更糟，于是同样闭嘴。
 *
 * 路径按**绑定根**相对返回（与 list / grep / index_files 一致，服务端
 * `LocalWorkspace._out` 再剥掉子路径前缀）；旁路区与忽略名单则按 `cwdAbs`（本工作区的
 * 有效根）判定，因为 `AgentCore/` 旁路区就挂在那一层。
 */
export async function withWrittenFiles(
  result: WorkspaceOpResult,
  opts: { rootAbs: string; cwdAbs: string; cutoffMs: number },
): Promise<WorkspaceOpResult> {
  if (!result.ok || typeof result.value !== "object" || result.value === null) {
    return result;
  }
  const value = result.value as Record<string, unknown>;
  if (!completedOnItsOwn(value)) return result;
  const scan = await scanWrittenFiles(opts.cwdAbs, opts.cutoffMs);
  const basePrefix = toPosix(relative(opts.rootAbs, opts.cwdAbs));
  const prefix = basePrefix && basePrefix !== "." ? `${basePrefix}/` : "";
  return {
    ok: true,
    value: { ...value, written_files: scan.files.map((p) => `${prefix}${p}`) },
  };
}

File: apps/server/agentcore/tools/sandbox/written_scan.py
"""产物写回 · 直写沙箱的本地腿：扫描「本次执行改了工作区哪些文件」。

云桌 guest 与 ``SubprocessSandbox`` / 桌面 Local 都以工作区真盘为 cwd 写文件
（无 copy-out 枚举）。不事后看盘就没人能如实填 ``ExecutionResult.written_files``。
本模块是这条腿。

为什么是「执行后单次有界扫描 + mtime 截止」，而不是执行前后各拍一次全树快照：

* **成本几乎全在目录枚举次数上**，与文件数关系不大。实测本仓（剪掉 node_modules /
  .git 等噪音目录后仍有 19.5k 目录 / 196k 文件）：整树带 stat ~940ms，只枚举目录不
  stat 文件也要 ~870ms。前后两次快照 = 双倍这个代价，会把一条 100ms 的脚本拖成两秒。
  单次事后扫描省掉一半。
* **有界**：从 cwd 起 BFS，撞到 :data:`_MAX_DIRS` 或 :data:`_BUDGET_MS` 立刻收手。
  BFS 保证被截断的永远是最深的尾巴，而产物几乎都落在 cwd 或浅层子目录；于是「超大
  工作区」退化成「浅层如实、极深处可能漏报」，而不是每次执行整体变慢。代价边界：
  单次执行最多多花 ~``_BUDGET_MS``（普通工作区通常个位数毫秒，实测 4 层以内 6ms）。
* **代价**：mtime 截止有一条边界模糊带，见 :data:`_MTIME_MARGIN_NS`。

排除面（复用 ``_paths`` 两档忽略规则，不另起第三份名单）：路径感知旁路区
``AgentCore/{index,trash,baselines}``、系统噪音目录 ``IGNORED_DIRS``、系统噪音后缀
（``*.db`` / ``*.pyc``）、符号链接。**不**排除 AI 噪音后缀——AI 生成的 ``.png`` 图表、
``.zip`` 打包件正是交付物。

扫描连同 ``IGNORED_DIRS`` 一起剪掉（含 ``node_modules`` / ``out`` / ``dist``）。
三条理由——枚举这些目录正是成本大头；它们对 AI 视图与**用户文件面板**都隐藏；
名单共用单一真源。文件仍在真盘上，只是不进 ``written_files`` 台账。
"""

from __future__ import annotations

import os
import time
from collections import deque
from dataclasses import dataclass
from pathlib import Path

from agentcore.workspace._paths import (
    is_ignored_dir_name,
    is_internal_zone_relpath,
    is_system_ignored_file_name,
)

# 目录枚举上限。实测本仓 ~45µs/目录，4000 目录 ≈ 180ms，与下面的墙钟预算同量级。
_MAX_DIRS = 4000
# 墙钟兜底：网络盘 / 被杀毒软件挂钩的目录上，单次 scandir 可能慢几个数量级，
# 目录数上限就兜不住了。两道闸都要。
_BUDGET_MS = 250.0
# 报告条数上限：一次写出上千个文件（构建产物）时，列全既冲爆模型上下文也没有信息量。
_MAX_FILES = 200

# 文件系统时间戳粒度比墙钟粗且**向下截断**：刚起步就写的文件，mtime 可能落在我们读到
# 的启动时刻之前一点。实测 Windows/NTFS ~1ms；Linux 走粗粒度时钟（1 tick，HZ=100 时
# 10ms）；macOS/APFS ~0。截止值往前让 100ms 即十倍余量——而且解释器自身启动就要几十
# 毫秒，真正的写盘远在截止值之后，这条余量基本只是保险。
#
# 为什么不让更多：余量就是误报窗口（窗口内**别人**刚改过的文件会被算进本次执行）。
# 已知不覆盖的两种盘，宁可如实说明也不用秒级余量把所有人的精度赔进去：
# FAT/exFAT 的修改时间是 2 秒粒度（开头 ~2s 内的产物可能漏报）；网络盘（NFS/SMB）
# 若服务端时钟落后于本机，任何余量都救不回来。执行前就存在且未变动的文件在哪种盘上
# 都不会命中——它们的 mtime 是旧的。
_MTIME_MARGIN_NS = 100_000_000


@dataclass(frozen=True)
class WrittenScanResult:
    """一次事后扫描的结果。``truncated`` = 撞了预算，深处可能还有没看到的改动。"""

    files: list[str]
    truncated: bool = False


def written_scan_cutoff_ns() -> int:
    """执行开始前取的 mtime 截止值（已含 :data:`_MTIME_MARGIN_NS` 余量）。"""
    return time.time_ns() - _MTIME_MARGIN_NS


def scan_written_files(
    root: Path,
    *,
    cutoff_ns: int,
    max_dirs: int = _MAX_DIRS,
    budget_ms: float = _BUDGET_MS,
    max_files: int = _MAX_FILES,
) -> WrittenScanResult:
    """``root`` 下 mtime 不早于 ``cutoff_ns`` 的常规文件（工作区相对 POSIX 路径）。

    同步阻塞（大量 ``scandir``），调用方须放到线程里跑。任何单条目的 ``OSError``
    （权限 / 占用 / 竞态删除）只跳过该条目，绝不让整次扫描失败——执行本身已经成功了。
    """
    deadline = time.monotonic() + budget_ms / 1000.0
    queue: deque[tuple[str, str]] = deque([("", str(root))])
    found: list[str] = []
    dirs_seen = 0
    truncated = False

    while queue:
        if dirs_seen >= max_dirs or time.monotonic() >= deadline:
            truncated = True
            break
        parent_rel, abs_dir = queue.popleft()
        dirs_seen += 1
        try:
            entries = list(os.scandir(abs_dir))
        except OSError:
            continue
        for entry in entries:
            child_rel = f"{parent_rel}/{entry.name}" if parent_rel else entry.name
            try:
                # follow_symlinks=False 两处都给：符号链接既不入队也不上报。
                if entry.is_dir(follow_symlinks=False):
                    # 等价于 ``is_ignored_dir_entry``：既然从不下潜进噪音目录，
                    # 它那趟「祖先段是否噪音」的复检在这里恒为假，省掉。
                    if is_ignored_dir_name(entry.name) or is_internal_zone_relpath(
                        child_rel
                    ):
                        continue
                    queue.append((child_rel, entry.path))
                    continue
                if not entry.is_file(follow_symlinks=False):
                    continue
                if is_system_ignored_file_name(entry.name):
                    continue
                if entry.stat(follow_symlinks=False).st_mtime_ns >= cutoff_ns:
                    found.append(child_rel)
            except OSError:
                continue
        if len(found) >= max_files:
            truncated = True
            break

    found.sort()
    return WrittenScanResult(files=found[:max_files], truncated=truncated)
